plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,182 @@
+    """
+    try:
+      dts = '{0:d}-{1:d}-{2:d} {3:d}:{4:d}:{5:d}.{6:d} {7:s}'.format(
+          structure['year'], structure['month'], structure['day_of_month'],
+          structure['time']['hours'], structure['time']['minutes'],
+          structure['time']['seconds'], structure.get('microseconds', [0])[0],
+          structure['time_zone'])
+    except (TypeError, ValueError) as exception:
+      raise ValueError(
+          'unable to format date time string with error: {0!s}.'.format(
+              exception))
+    return parser.parse(timestr=dts)
+
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a record and produces a PostgreSQL event.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): elements parsed from the file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key != 'logline':
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    event_data = PostgreSQLEventData()
+    event_data.pid = ''.join(
+        [str(x) for x in self._GetValueFromStructure(structure, 'pid')])
+
+    log_level = self._GetValueFromStructure(structure, 'log_level')
+    if log_level and len(log_level) == 1:
+      event_data.log_level = log_level[0]
+    else:
+      parser_mediator.ProduceExtractionWarning('no log level found')
+      return
+
+    user_and_database = self._GetValueFromStructure(
+        structure, 'user_and_database')
+    if user_and_database:
+      event_data.user = ''.join(user_and_database)
+
+    log_line = self._GetValueFromStructure(structure, 'log_line')
+    if log_line:
+      event_data.log_line = log_line.lstrip().rstrip()
+
+    date_time_structure = self._GetValueFromStructure(structure, 'date_time')
+    date_time_parsed = self._ConvertTimeString(date_time_structure)
+    date_time = dfdatetime_time_elements.TimeElementsInMicroseconds()
+    date_time.CopyFromDatetime(date_time_parsed)
+    event = time_events.DateTimeValuesEvent(
+        date_time, definitions.TIME_DESCRIPTION_RECORDED)
+    parser_mediator.ProduceEventWithEventData(event, event_data)
+
+  # pylint: disable=unused-argument
+  def VerifyStructure(self, parser_mediator, lines):
+    """Verifies that this is a bash history file.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between
+          parsers and other components, such as storage and dfvfs.
+      lines (str): one or more lines from the text file.
+
+    Returns:
+      bool: True if this is the correct parser, False otherwise.
+    """
+    try:
+      self._LOG_LINE.parseString(lines)
+    except pyparsing.ParseException as exception:
+      logger.debug('Not a PostgreSQL log file: {0!s}'.format(exception))
```

Comment:
no need for this debug log line, please remove

Reply:
Done